fdbrpc/libeio/eio.c

```diff
@@ -2217,7 +2217,9 @@ X_THREAD_PROC (etp_proc)
 
       X_LOCK (reslock);
 
+      X_LOCK(reqlock);
       ++npending;
+      X_UNLOCK(reqlock);
```

Comment:
I'm a little bit unusre about this one, because almost all mutations of `npending`, including this one, happen in the `reslock`. The only place that uses `reqlock` is the function that reads this variable (`etp_npending`). I'm also slightly nervous about acquiring this lock inside a critical section for `reslock` without carefully looking over the rest of the code.

It feels like we should either use `reslock` for `npending` and update `etp_npending`, or update all of the places that access `npending` to use `reqlock`. If the latter, we should consider moving the increment out of `reslock` and into `reqlock`. In both cases, we would probably need to make sure it's valid to do so.

It might be possible for us to inspect more recent versions of this source and see if and how it was fixed.

Reply:
I think I found the most recent source here: http://cvs.schmorp.de/libeio/etp.c?revision=1.14&view=markup#l398. Doesn't look like it's changed.

What would you think of using atomics? It seems like the original author was under the impression that writing and reading a word concurrently is safe, so we could just make that the case.

Reply:
Sounds fine to me